### PR TITLE
feat(debugger): Print limited source code context

### PR DESCRIPTION
**Pull request by mverzilli**
# Description

## Problem

This PR limits the amount of lines of source code printed after each debug step. Before this change, the debugger just displayed the whole source file after each step, which is inconvenient when working with medium and larger source files. 

Part of #3015 .

## Summary

Besides the implementation of the behavior itself, some notes about things  to the reviewer.

- Adds a dependency to the codespan-reporting crate.
- Adds a bunch of functions to DebugArtifact, in order to query it given a Location.
- Re-exports the noirc_errors::Location type from nargo::errors.

## Additional Context

Demo of the new behavior. 

limit-context

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: tooling/debugger/src/lib.rs [hunk 1]
```diff
@@ -74,25 +79,77 @@ impl<'backend, B: BlackBoxFunctionSolver> DebugContext<'backend, B> {
...
     }
 
-    fn show_source_code_location(location: &OpcodeLocation, debug_artifact: &DebugArtifact) {
+    fn print_location_path(&self, loc: Location) {
+        let line_number = self.debug_artifact.location_line_number(loc).unwrap();
+        let column_number = self.debug_artifact.location_column_number(loc).unwrap();
+
+        println!(
+            "At {}:{line_number}:{column_number}",
+            Files::name(&self.debug_artifact, loc.file).unwrap()
```
Comment (TomAFrench) [inline tooling/debugger/src/lib.rs:92]:
```suggestion
            self.debug_artifact.name(loc.file).unwrap()
```

File: tooling/debugger/src/lib.rs [hunk 2]
```diff
@@ -74,25 +79,77 @@ impl<'backend, B: BlackBoxFunctionSolver> DebugContext<'backend, B> {
...
+
+                let first_line_to_print =
+                    if loc_line_index < context_lines { 0 } else { loc_line_index - context_lines };
+
+                let last_line_index = debug_artifact.last_line_index(loc).unwrap();
+                let last_line_to_print = if loc_line_index + context_lines > last_line_index {
+                    last_line_index
+                } else {
+                    loc_line_index + context_lines
+                };
```
Comment (TomAFrench) [inline tooling/debugger/src/lib.rs:116]:
```suggestion
                let last_line_to_print = std::cmp::min(loc_line_index + context_lines, last_line_index);
```

File: tooling/debugger/src/lib.rs [hunk 3]
```diff
@@ -74,25 +79,77 @@ impl<'backend, B: BlackBoxFunctionSolver> DebugContext<'backend, B> {
...
+                let source = debug_artifact.location_source_code(loc).unwrap();
+                for (current_line_index, line) in source.lines().enumerate() {
+                    let current_line_number = current_line_index + 1;
+
+                    if current_line_index < first_line_to_print {
+                        // Ignore lines before range starts
+                        continue;
+                    } else if current_line_index == first_line_to_print && current_line_index > 0 {
+                        // Denote that there's more lines before but we're not showing them
+                        print_ellipsized_line(current_line_index);
```
Comment (TomAFrench) [inline tooling/debugger/src/lib.rs:127]:
```suggestion
                        print_line_of_ellipses(current_line_index);
```

"ellipsized" feels a bit weird to me.

File: tooling/nargo/src/artifacts/debug.rs [hunk 4]
```diff
@@ -45,6 +45,53 @@ impl DebugArtifact {
 
         Self { debug_symbols, file_map }
     }
+
+    /// Given a location, returns its file's source code
+    pub fn location_source_code(&self, location: Location) -> Result<&str, Error> {
+        Files::source(self, location.file)
```
Comment (TomAFrench) [inline tooling/nargo/src/artifacts/debug.rs:51]: We don't tend to use the `Trait::method(self,...)` syntax, could you change this to `self.method(...)` instead please?

File: tooling/nargo/src/artifacts/debug.rs [hunk 5]
```diff
@@ -45,6 +45,53 @@ impl DebugArtifact {
...
+    /// Given a location, returns the index of the line it starts at
+    pub fn location_line_index(&self, location: Location) -> Result<usize, Error> {
+        let location_start = location.span.start() as usize;
+        Files::line_index(self, location.file, location_start)
+    }
+
+    /// Given a location, returns the line number it starts at
+    pub fn location_line_number(&self, location: Location) -> Result<usize, Error> {
+        let location_start = location.span.start() as usize;
+        let line_index = Files::line_index(self, location.file, location_start).unwrap();
```
Comment (TomAFrench) [inline tooling/nargo/src/artifacts/debug.rs:63]: We shouldn't need to unwrap these errors as we can just `?` them.

File: tooling/debugger/src/lib.rs [hunk 6]
```diff
@@ -74,25 +79,77 @@ impl<'backend, B: BlackBoxFunctionSolver> DebugContext<'backend, B> {
...
+
+        println!(
+            "At {}:{line_number}:{column_number}",
+            Files::name(&self.debug_artifact, loc.file).unwrap()
+        );
+    }
+
+    fn show_source_code_location(&self, location: &OpcodeLocation, debug_artifact: &DebugArtifact) {
         let locations = debug_artifact.debug_symbols[0].opcode_location(location);
         if let Some(locations) = locations {
```
Comment (TomAFrench) [inline tooling/debugger/src/lib.rs:98]:
```suggestion
        let Some(locations) = locations else { return }
```

This will help remove some indentation.

**Review by TomAFrench — COMMENTED** [review]
Tested this locally and it looks good to me. Review basically boils down to some nits on code style.

I'm also getting multiple lines printed for each opcode (the entire depth of the stack trace from `main`?). We may want to limit this in future for deep programs but it's not a blocker.

File: tooling/debugger/src/lib.rs (hunk 3 above)
Comment (mverzilli) [reply tooling/debugger/src/lib.rs:127]: It's certainly an abuse of language 😅 . Will change it.

File: tooling/debugger/src/lib.rs (hunk 6 above)
Comment (mverzilli) [reply tooling/debugger/src/lib.rs:98]: Neat 👌

**Review by TomAFrench — APPROVED** [review]

**Pull request merged**